runtime/src/accounts_index.rs

```diff
@@ -204,30 +204,94 @@ impl<T: 'static + Clone + IsCached> WriteAccountMapEntry<T> {
         })
     }
 
+    fn addref(item: &AtomicU64) {
+        item.fetch_add(1, Ordering::Relaxed);
+    }
+
+    pub fn upsert_new_key<'a>(
+        mut w_account_maps: AccountMapsWriteLock<'a, T>,
+        pubkey: &Pubkey,
+        new_value: AccountMapEntry<T>,
+        reclaims: &mut SlotList<T>,
+    ) {
+        match w_account_maps.entry(*pubkey) {
+            Entry::Occupied(mut occupied) => {
+                let current = occupied.get_mut();
+                Self::upsert_item(current, new_value, reclaims);
+            }
+            Entry::Vacant(vacant) => {
+                vacant.insert(new_value);
+            }
+        }
+    }
+
+    // returns None if upsert was successful
+    // otherwise, the pubkey has to be added to the maps with a write lock. call upsert_new
+    pub fn upsert_existing_key<'a>(
```

Comment:
nit: upsert implies creation of a new key if it doesn't exist, maybe rename to `update_if_exists()`